T7122: pki: unable to switch from custom cert to ACME when HAProxy service is running with 'redirect-http-to-https' option

c-po opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->

HAProxy option `redirect-http-to-https` will block port 80 used by certbot.

When requesting certificates via ACME/certbot we will check if the requested certificate will be in use by a conflicting service - currently only HAProxy . If this is the case certbot will only listen on localhost:65080 and will require the connection be proxied via HAProxy .

HAProxy will check if its frontend service certificate is ACME based and if so - automatically render the ACL rules to reverse-proxy the requests down to certbot via localhost:65080.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
* https://vyos.dev/T7122

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Manual Testcases

### Moving from embedded certificate to ACME :heavy_check_mark:

This testcase is the original one triggering the bug. We're using an embedded certificate for `load-balancing haproxy` - embedded means base64 encoded certificate data provided via CLI.

```
set pki certificate cloud.XXX.net certificate 'MIIEMj...'
set pki certificate cloud.XXX.net private key 'MIIEMj...'
set load-balancing haproxy service cloud ssl certificate 'cloud.XXX.net'
```

Which is about to be replaced with a one issued via ACME (in this case the LetsEncrypt Staging API)

```
set pki certificate LE_cloud acme domain-name cloud.XXX.net
set pki certificate LE_cloud acme email ca@XXX.net
set pki certificate LE_cloud acme url https://acme-staging-v02.api.letsencrypt.org/directory
set load-balancing haproxy service cloud redirect-http-to-https
```

The issue in this test-case is that port 80 is already in use by HAProxy due to `redirect-http-to-http` - we need to reverse-proxy
the request for the certificate.

Before:
```bash
$ openssl s_client -connect cloud.XXX.net:443 </dev/null 2>/dev/null | openssl x509 -noout -issuer -dates -subject
issuer=O = CAcert Inc., OU = http://www.CAcert.org, CN = CAcert Class 3 Root
notBefore=Sep  2 17:33:31 2023 GMT
notAfter=Sep  1 17:33:31 2025 GMT
subject=CN = cloud.XXX.net
```

Now change the certificate to an ACME generated one:

```
vyos@vyos# compare
[load-balancing haproxy service frontend ssl]
- certificate "ipsec"
+ certificate "LE_cloud"
[pki certificate]
+ LE_cloud {
+     acme {
+         domain-name "cloud.XXX.net"
+         email "ca@XXX.net"
+         url "https://acme-staging-v02.api.letsencrypt.org/directory"
+     }
+ }

vyos@vyos# commit
[ pki ]
PKI: Updating config: load-balancing haproxy service frontend ssl certificate LE_cloud
Adding/replacing automatically imported CA certificate for "LE_cloud" ...
```

After:

```bash
$ openssl s_client -connect cloud.XXX.net:443 </dev/null 2>/dev/null | openssl x509 -noout -issuer -dates -subject
issuer=C = US, O = (STAGING) Let's Encrypt, CN = (STAGING) Counterfeit Cashew R10
notBefore=May  4 08:59:04 2025 GMT
notAfter=Aug  2 08:59:03 2025 GMT
subject=CN = cloud.XXX.net
```

###  Bootstrapping :heavy_check_mark:

There is neither a running HAProxy instance, nor any ACME issued certificate. This must be detected during initial bootstrapping and handled later for cert renewal.

```
set pki certificate LE_cloud acme domain-name 'cloud.XXX.net'
set pki certificate LE_cloud acme email 'ca@XXX.net'
set pki certificate LE_cloud acme url 'https://acme-staging-v02.api.letsencrypt.org/directory'

set load-balancing haproxy backend atlas server atlas address '192.0.2.1'
set load-balancing haproxy backend atlas server atlas port '443'
set load-balancing haproxy backend atlas ssl no-verify
set load-balancing haproxy service frontend backend 'atlas'
set load-balancing haproxy service frontend port '443'
set load-balancing haproxy service frontend redirect-http-to-https
set load-balancing haproxy service frontend ssl certificate 'LE_cloud'
```

```
vyos@vyos# commit
[ pki ]

Updating configuration: "load-balancing haproxy service frontend ssl certificate LE_cloud"

Add/replace automatically imported CA certificate for "LE_cloud"...

Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
```

```bash
# openssl s_client -connect cloud.XXX.net:443 </dev/null 2>/dev/null | openssl x509 -noout -issuer -dates -subject
issuer=C = US, O = (STAGING) Let's Encrypt, CN = (STAGING) Wannabe Watercress R11
notBefore=May  4 19:23:11 2025 GMT
notAfter=Aug  2 19:23:10 2025 GMT
subject=CN = cloud.XXX.net
```

### Deprovisioning :heavy_check_mark:

Remove ACME issued certificate from `load-balancing haproxy` and return to an embedded certificate.

```bash
$ openssl s_client -connect cloud.XXX.net:443 </dev/null 2>/dev/null | openssl x509 -noout -issuer -dates -subject
issuer=O = CAcert Inc., OU = http://www.CAcert.org, CN = CAcert Class 3 Root
notBefore=Sep  2 17:33:31 2023 GMT
notAfter=Sep  1 17:33:31 2025 GMT
subject=CN = cloud.XXX.net
```

### Deleting all CLI nodes for either PKI, HAProxy or both :heavy_check_mark:

```
vyos@vyos# delete pki
vyos@vyos# delete load-balancing
vyos@vyos# commit
[ pki ]
Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
```

```bash
$ openssl s_client -connect cloud.XXX.net:443 </dev/null 2>/dev/null | openssl x509 -noout -issuer -dates -subject
Could not read certificate from <stdin>
Unable to load certificate
```

## Smoketests :heavy_check_mark:

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_load-balancing_haproxy.py
test_reverse_proxy_backend_http_check (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_backend_http_check) ... ok
test_reverse_proxy_backend_ssl_no_verify (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_backend_ssl_no_verify) ... ok
test_reverse_proxy_ca_not_exists (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_ca_not_exists) ... ok
test_reverse_proxy_cert_not_exists (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_cert_not_exists) ... ok
test_reverse_proxy_domain (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_domain) ... ok
test_reverse_proxy_http_compression (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_compression) ... ok
test_reverse_proxy_http_redirect (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_redirect) ... ok
test_reverse_proxy_http_response_headers (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_response_headers) ... ok
test_reverse_proxy_logging (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_logging) ... ok
test_reverse_proxy_tcp_health_checks (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_tcp_health_checks) ... ok
test_reverse_proxy_tcp_mode (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_tcp_mode) ... ok
test_reverse_proxy_timeout (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_timeout) ... ok

----------------------------------------------------------------------
Ran 12 tests in 147.001s

OK
```

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_pki.py
test_certificate_eapol_update (__main__.TestPKI.test_certificate_eapol_update) ... ok
test_certificate_https_update (__main__.TestPKI.test_certificate_https_update) ... ok
test_certificate_in_use (__main__.TestPKI.test_certificate_in_use) ... ok
test_invalid_ca_valid_certificate (__main__.TestPKI.test_invalid_ca_valid_certificate) ... ok
test_valid_pki (__main__.TestPKI.test_valid_pki) ... ok

----------------------------------------------------------------------
Ran 5 tests in 62.816s

OK
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

c-po committed:
vyos.utils: T7122: fix IPv6 support in check_port_availability()

Commit 4523e9c897b3 ("wireguard: T3763: Added check for listening port
availability") added a function to check if a port is free to use or already
occupied by a different running service. This has been done by trying to bind a
socket to said given port.

Unfortunately there is no support for IPv6 address-fdamily in both
socketserver.TCPServer or socketserver.UDPServer. This must be done manually by
deriving TCPServer and setting self.address_family for IPv6.

The new implementation gets rid of both TCPServer and UDPServer and replaces it
with a simple socket binding to a given IPv4/IPv6 address or any interface/
address if unspecified.

In addition build time tests are added for the function to check for proper
behavior during build time of vyos-1x.

c-po committed:
haproxy: T7122: render explicit http configuration to properly bind port 80

If redirect-http-to-https is set we will render a discrete onfiguration in
HAproxy to properly claim port 80 in the system to detect if a service is
alreadey using the port or not.

c-po committed:
T7122: remove trailing chars and add new line for every template.render() call

c-po committed:
pki: T7122: remove duplicate list definition - can be auto generated

changed_keys had the same content as the values inside the sync_translate
dictionary. Infact they were both used together do defined changed CLI keys.
The list for changed_keys is a list of all unique values inside the
sync_translate dict.

c-po committed:
pki: T7122: place certbot behind reverse-proxy if cert used by haproxy

If we detect that an ACME issued certificate is consumed by haproxy service,
we will move the certbot webserver to localhost and a highport, to proxy the
request via haproxy which is already using port 80.

c-po committed:
haproxy: T7122: do not use f'ormat strings without variable

c-po committed:
haproxy: T7122: automatically reverse-proxy to certbot

Automatically render HaProxy rules to reverse-proxy ACME challanges when the
requested certificate was issued using ACME.

c-po committed:
op-mode: T7122: add "show|monitor log haproxy" commands

c-po committed:
xml: T7122: it is spelled HAProxy

c-po committed:
vyos.template: T7122: add Jinja2 clever function helper to read vyos.defaults

Add a new category if Jinja2 operands. We already have filters and tests, but
sometimes we would like to call a Python function without and data "|" piped
to it - that's what they call a clever-function.

{{ get_default_port(NAME) }} can be used to retrieve the value from
vyos.defaults.internal_ports[NAME] within Jinja2. We no longer need to extend
the dictionary with arbitrary data retrieved from vyos.defaults, we can now
simply register another clever-function to the Jinja2 backend.

c-po committed:
haproxy: T7122: always reverse-proxy ACL for certbot

Always enable the ACL entry to reverse-proxy requests to the path
"/.well-known/acme-challenge/" when "redirect-http-to-https" is configured for
a given HAProxy frontend service.

This is an intentional design decision to simplify the implementation and reduce
overall code complexity. It poses no risk: a missing path returns a 404, and an
unavailable backend yields an error 503.

This approach avoids a chicken-and-egg problem where certbot might try to
request a certificate via reverse-proxy before the proxy config is actually
generated and active.

By always routing through HAProxy, we also eliminate downtime as port 80 does
not need to be freed for certbot's standalone mode.

c-po committed:
vyos.base: T7122: add new Message() helper wrapper for print()

This will wrap the messages at 72 characters in the same way as Warning() and
DeprecationWarning() would do. We now have simple wrappers for it!

Example:
vyos@vyos# commit
[ pki ]

Updating configuration: "load-balancing haproxy service frontend ssl
certificate LE_cloud"

Add/replace automatically imported CA certificate for "LE_cloud"

c-po committed:
pki: T7122: extend ca/certificate removal check to lists

Some VyOS CLI nodes support defining multiple certificates. The previous check
when removing a certificate from the CLI only performed a string comparison,
which failed in cases where the underlying data was a list (CLI <multi/> node).

This update extends the check to handle both cases:
- If the datum is a string, perform a string comparison.
- If the datum is a list, check whether the target certificate is part of the
  list.

This ensures proper removal behavior regardless of the data type used in the
CLI node.

c-po committed:
haproxy: T7122: add ACME/certbot bootstrap support

When both the CLI PKI node for an ACME-issued certificate and HAProxy are
configured during initial setup, the certbot challenge cannot be served via the
reverse proxy because HAProxy has not yet been configured at all.

This commit introduces a special case to handle this bootstrap scenario,
ensuring that the certbot challenge can still be served correctly in standalone
mode on port 80 despite initial config dependencies/priorities between PKI
and HAProxy.

sever-sever commented on python/vyos/template.py:
```diff
@@ -1050,3 +1074,21 @@ def vyos_defined(value, test_value=None, var_type=None):
+    """
+    from vyos.defaults import internal_ports
+    if service not in internal_ports:
+        raise RuntimeError(f'Service "{service}" not found in internal ' \
```
Unnecessary backslash :)

sever-sever reviewed (approved):
Big change, covered by smoke test. Do not have anything against, though I cannot check all manually.

Fixes validators for address usage for IPv6
Fixes ACME changes while haproxy binds port 80
Improved error messages/messages

c-po committed:
pki: T7122: when ACME listen-address is used - check if port is available

When instructing certbot to listen on a given address, check if the address is
free to use. Also take this into account when spawning certbot behind HAProxy.
If the address is not (yet) bound - the request must be done in standalone mode
and not via the reverse-proxy.

sarthurdev commented:
Issue in smoketest

```
DEBUG - ======================================================================
DEBUG - FAIL: test_reverse_proxy_http_redirect (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_redirect)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_load-balancing_haproxy.py", line 606, in test_reverse_proxy_http_redirect
DEBUG -     self.assertIn(f'use_backend certbot_{haproxy_service_name}_backend if acme_acl', config[frontend_name])
DEBUG - AssertionError: 'use_backend certbot_https_front_backend if acme_acl' not found in 'mode http\nbind [::]:80 v4v6\nacl acme_acl path_beg /.well-known/acme-challenge/\nuse_backend buildin_acme_certbot if acme_acl\nredirect scheme https code 301 if !acme_acl'
DEBUG - 
DEBUG - ----------------------------------------------------------------------
DEBUG - Ran 12 tests in 72.520s
```

c-po commented:
> Issue in smoketest
> 
> ```
> DEBUG - ======================================================================
> DEBUG - FAIL: test_reverse_proxy_http_redirect (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_redirect)
> DEBUG - ----------------------------------------------------------------------
> DEBUG - Traceback (most recent call last):
> DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_load-balancing_haproxy.py", line 606, in test_reverse_proxy_http_redirect
> DEBUG -     self.assertIn(f'use_backend certbot_{haproxy_service_name}_backend if acme_acl', config[frontend_name])
> DEBUG - AssertionError: 'use_backend certbot_https_front_backend if acme_acl' not found in 'mode http\nbind [::]:80 v4v6\nacl acme_acl path_beg /.well-known/acme-challenge/\nuse_backend buildin_acme_certbot if acme_acl\nredirect scheme https code 301 if !acme_acl'
> DEBUG - 
> DEBUG - ----------------------------------------------------------------------
> DEBUG - Ran 12 tests in 72.520s
> ```

Fixed 

```
cpo@LR5.wue4:~$ /usr/libexec/vyos/tests/smoke/cli/test_load-balancing_haproxy.py
test_reverse_proxy_backend_http_check (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_backend_http_check) ... ok
test_reverse_proxy_backend_ssl_no_verify (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_backend_ssl_no_verify) ... ok
test_reverse_proxy_ca_not_exists (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_ca_not_exists) ... ok
test_reverse_proxy_cert_not_exists (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_cert_not_exists) ... ok
test_reverse_proxy_domain (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_domain) ... ok
test_reverse_proxy_http_compression (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_compression) ... ok
test_reverse_proxy_http_redirect (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_redirect) ... ok
test_reverse_proxy_http_response_headers (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_http_response_headers) ... ok
test_reverse_proxy_logging (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_logging) ... ok
test_reverse_proxy_tcp_health_checks (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_tcp_health_checks) ... ok
test_reverse_proxy_tcp_mode (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_tcp_mode) ... ok
test_reverse_proxy_timeout (__main__.TestLoadBalancingReverseProxy.test_reverse_proxy_timeout) ... ok

----------------------------------------------------------------------
Ran 12 tests in 274.214s

OK
```

dmbaturin reviewed (approved):
The logic is sensible and the tests are fixed, no objections.
